### PR TITLE
Add support for NHWC GridSample in the CUDA EP and enable grid_sample_test for all EPs

### PR DESCRIPTION
**Pull request by mtavenrath**
### Description
I've added NHWC GridSample support to the CUDA EP to reduce the number of layout transforms. Also I've enabled the full set of GridSampleTests for all EPs. I've also added the GridSample OpSet 16 to the registered kernels.

### Motivation and Context
This is the first PR is a series of enhancements of the CUDA EP improving NHWC support to avoid costly layout transforms between NWHC and NCHW nodes which are layout sensitive. Also testing was quite rudimentary for the CUDA EP while it was great for the CPU path. I've regenerated grid_sample_test.cc enabling tests for other platforms as well. Those tests resurfaced #10607 again which is fixed as well.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/cuda_provider_factory.cc [hunk 1]
```diff
@@ -219,6 +219,9 @@ struct CUDA_Provider : Provider {
     info.cudnn_conv_use_max_workspace = params->cudnn_conv_use_max_workspace != 0;
     info.enable_cuda_graph = params->enable_cuda_graph != 0;
     info.prefer_nhwc = params->prefer_nhwc;
+    // HACK
+    info.prefer_nhwc = true;
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/cuda_provider_factory.cc:223]: revert this change

File: onnxruntime/contrib_ops/cuda/grid_sample.cc [hunk 2]
```diff
@@ -9,22 +9,23 @@ namespace onnxruntime {
...
 namespace cuda {
 
-#define REGISTER_KERNEL_TYPED(T)                                   \
+#define REGISTER_KERNEL_TYPED(T, VERSION, LAYOUT, DOMAIN)                   \
   ONNX_OPERATOR_TYPED_KERNEL_EX(                                   \
       GridSample,                                                  \
-      kMSDomain,                                                   \
-      1,                                                           \
+      DOMAIN,                                                      \
+      VERSION,                                                          \
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/grid_sample.cc:16]: nit: Format code with Clang-format like https://skillupwards.com/blog/formatting-c-cplusplus-code-using-clangformat-and-vscode.

Comment (mtavenrath) [reply onnxruntime/contrib_ops/cuda/grid_sample.cc]: Done.

File: onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc [hunk 3]
```diff
@@ -25,8 +34,7 @@ TEST(GridsampleTest, test_grid_sample_16_4D_nearest_zeros_align_corners) {
   test.AddAttribute("padding_mode", padding_mode);
   test.AddAttribute("align_corners", align_corners);
   test.AddOutput<float>("Y", Y_shape, Y_data);
-  test.ConfigEp(DefaultCpuExecutionProvider())
-      .RunWithConfig();
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "", GetExcludedExecutionProviders(16));
```
Comment (tianleiwu) [inline onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc:37]: Suggest to use ConfigEps like the following:
```
  std::vector<std::unique_ptr<IExecutionProvider>> execution_providers;
  execution_providers.emplace_back(DefaultCpuExecutionProvider());
#ifdef USE_CUDA
  execution_providers.emplace_back(DefaultCudaExecutionProvider());
#endif
  test.ConfigEps(std::move(execution_providers)).RunWithConfig();
 ```
 
Need change the test to cover both NCHW and NHWC: 
if ENABLE_CUDA_NHWC_OPS is defined, test NHWC like the following: transpose X, and Y, change operator domain to internal NHWC domain, then run the test again.

Comment (mtavenrath) [reply onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc:37]: This change already improves GridSample testing a lot on all EPs and even found one bug in the current CUDA implementation. 

To get full NHWC testing without having to modify all tests the best idea I have come up so far is add a new Cuda NHWC EP which sets prefer_nwhc = true in the factory. This way all tests can run with the CUDA EP and the CUDA NHWC EP automatically. I haven't verified yet if this will break any CUDA EP based special handling.

Comment (mtavenrath) [reply onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc:37]: With regards to your proposal, before I tested the same idea by passing the `execution_providers` vector to `BaseTester::Run()` with the result that only the first EP in the list is executed.

The same thing happens with your proposal passing a vector of execution providers to the `ConfigEps` function. If this is supposed to work I'll file an issue.

To get this PR integrated my proposal is that I'll add all known EPs except for the CUDA EP and CPU EP into the excluded list.

Unfortunately it is not possible to have two CUDA EPs with two different configurations in the list for testing. Thus it's not possible to enable NHWC testing by default for now. I'll take a closer look if there is an easy workaround for this issue.

Comment (tianleiwu) [reply onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc:37]: You are right that only one provider is used in testing (the first one that has the implementation) so we add CUDA EP to the list first before CPU provider. In this way, CUDA provider is tested in cuda build, and CPU provider is tested in CPU build.

Another way is to test each provider separately like [this](https://github.com/microsoft/onnxruntime/blob/1b48054e1b7991ccef664fbedd659ec95d0e7ca7/onnxruntime/test/contrib_ops/attention_op_test.cc#L230-L252)

Comment (mtavenrath) [reply onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc:37]: I remember that I tried to test something like this as well and had problems with the failure report. I'll give it another try. Eventually it'd be good if the tester will handle all implementations for us to avoid code duplication.

The reason why I wanted to add the CUDA EP to this test is avoid code duplication. The NHWC tests are extremely incomplete and it doesn't make sense to copy/paste this test to the NHWC folder and the cuda test folder as well.

In my ideal world all EPs will get the same amount of testing to have the most consistent result.

Rewriting all test base class it out of scope for this CL.

My suggestion is that I'll remove the changed test and we run the same 'bad' tests as today for the GridSample implementation. In another change I'll address the issues of running on test with multiple EPs in a positive list and also allow running the same EP with different options to have the chance for even more complete testing.

Comment (tianleiwu) [reply onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc:37]: It's fine to use current change of this file to test NCHW cuda implementation.

My major concern is test coverage. Currently, this test does not cover NHWC. If you need test NHWC, need add domain like `OpTester test("GridSample", 16, kMSInternalNHWCDomain)`

The NHWC tests compares NCHW and NHWC results. I think we need at least one such test to ensure NHWC code is tested.

Comment (mtavenrath) [reply onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc]: I have implemented a small test function which run a test over a list of specified test. If probably could be added to the testing infrastructure in the future.

Test will run for NHWC & NCHW for the CUDA EP.

File: onnxruntime/test/util/default_providers.cc [hunk 4]
```diff
@@ -118,6 +121,17 @@ std::unique_ptr<IExecutionProvider> DefaultCudaExecutionProvider() {
   return nullptr;
 }
 
+std::unique_ptr<IExecutionProvider> DefaultCudaNHWCExecutionProvider() {
+#ifdef USE_CUDA
+  OrtCUDAProviderOptionsV2 provider_options{};
+  provider_options.do_copy_in_default_stream = true;
+  provider_options.prefer_nhwc = true;
```
Comment (tianleiwu) [inline onnxruntime/test/util/default_providers.cc:128]: prefer_nhwc=true option need ENABLE_CUDA_NHWC_OPS, see https://github.com/microsoft/onnxruntime/blob/57d6819212464f49b30db047528be0f409dadc67/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L245-L247

File: onnxruntime/test/util/default_providers.cc [hunk 5]
```diff
@@ -118,6 +121,17 @@ std::unique_ptr<IExecutionProvider> DefaultCudaExecutionProvider() {
   return nullptr;
 }
 
+std::unique_ptr<IExecutionProvider> DefaultCudaNHWCExecutionProvider() {
+#ifdef USE_CUDA
```
Comment (tianleiwu) [inline onnxruntime/test/util/default_providers.cc:125]: also require ENABLE_CUDA_NHWC_OPS  is defined.

File: onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc [hunk 6]
```diff
@@ -7,12 +7,30 @@
...
+#endif
+  return execution_providers;
+}
+
+template <typename T>
+void RunTests(T& test, std::vector<std::unique_ptr<IExecutionProvider>>& execution_providers) {
+  for (size_t idx = 0; idx < execution_providers.size(); ++idx) {
+    std::vector<std::unique_ptr<IExecutionProvider>> execution_provider;
+    execution_provider.push_back(std::move(execution_providers[idx]));
+    test.ConfigEps(std::move(execution_provider)).RunWithConfig();
```
Comment (tianleiwu) [inline onnxruntime/test/providers/cpu/tensor/grid_sample_test.cc:31]: nit: the code can be simplified as `test.ConfigEp(execution_providers[idx]).RunWithConfig();`

File: onnxruntime/test/util/default_providers.cc [hunk 7]
```diff
@@ -118,6 +121,17 @@ std::unique_ptr<IExecutionProvider> DefaultCudaExecutionProvider() {
   return nullptr;
 }
 
+std::unique_ptr<IExecutionProvider> DefaultCudaNHWCExecutionProvider() {
```
Comment (tianleiwu) [inline onnxruntime/test/util/default_providers.cc:124]: nit: Add `#ifdef ENABLE_CUDA_NHWC_OPS` for this function.

File: onnxruntime/test/util/default_providers.cc [hunk 8]
```diff
@@ -118,6 +121,17 @@ std::unique_ptr<IExecutionProvider> DefaultCudaExecutionProvider() {
...
+#ifdef USE_CUDA
+  OrtCUDAProviderOptionsV2 provider_options{};
```
[the remark this answers is not in the document]
Comment (mtavenrath) [reply onnxruntime/test/util/default_providers.cc:127]: This was my fault, I've protected the include for the struct by an nhwc ifdef , but didn't protect the function by the ifdef. fixed.

File: onnxruntime/contrib_ops/cuda/grid_sample_impl.cu [hunk 9]
```diff
@@ -239,9 +243,12 @@ void GridSampleImpl(
     T* output_data) {
   using Ch = Channels<IsNHWC>;
 
-  int blocksPerGrid = (int)(ceil(static_cast<T>(dims[Ch::N] * dims[Ch::C] * H_out * W_out) / GridDim::maxThreadsPerBlock));
+  int blocksPerGrid = static_cast<int>
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/grid_sample_impl.cu:246]: missing ( ) after static_cast<int>

**Comment by tianleiwu** [comment]
The kernelDocumentation pipeline failed, please download documents from
https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1301115&view=artifacts&pathAsName=false&type=publishedArtifacts

and update the files with same name under docs/

**Review by tianleiwu — APPROVED** [review]

File: onnxruntime/contrib_ops/cuda/grid_sample_impl.cu [hunk 10]
```diff
@@ -147,8 +169,9 @@ __global__ void _GridSampleKernel(
     if (grid_x_imgSpace < x_min || grid_x_imgSpace > x_max ||
         grid_y_imgSpace < y_min || grid_y_imgSpace > y_max) { // out of bound
       if (padding_mode == 1) {  // border
-        grid_x_imgSpace = max(0.0f, min(grid_x_imgSpace, W_in - 1.0f));
-        grid_y_imgSpace = max(0.0f, min(grid_y_imgSpace, H_in - 1.0f));
+        // Clamping must not be done here, see #10607
+        // grid_x_imgSpace = max(0.0f, min(grid_x_imgSpace, W_in - 1.0f));
```
Comment (hariharans29) [inline onnxruntime/contrib_ops/cuda/grid_sample_impl.cu:173]: Just curious: If there are tests that triggered this bug fixing (mentioned in the PR), how come the CPU version of the op not need similar fixing (it also has similar clamping) ?
